### PR TITLE
New package: quinoa

### PR DESCRIPTION
**Pull request by junghans**
Fixes quinoacomputing/quinoa#129

### CODE REVIEW

File: var/spack/repos/builtin/packages/quinoa/package.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    depends_on("netlib-lapack+lapacke")
+    depends_on("mad-numdiff")
+    depends_on("h5part")
+    depends_on("boostmplcartesianproduct")
+    depends_on("tut")
+    depends_on("pugixml")
+    depends_on("pstreams")
+    depends_on("pegtl")
+    
+
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/quinoa/package.py]: Two blank lines inside the class scope will cause flake8 to complain. You should use one blank line.

File: var/spack/repos/builtin/packages/quinoa/package.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+class Quinoa(CMakePackage):
+    """Quinoa is a set of computational tools that enables research and
+       numerical analysis in fluid dynamics. At this time it is a test-bed
+       to experiment with various algorithms using fully asynchronous runtime
+       systems.
+    """
+
+    homepage = "http://quinoacomputing.org"
+    url      = "https://github.com/quinoacomputing/quinoa/tarball/v1.0"
```
Comment (jbakosi) [inline var/spack/repos/builtin/packages/quinoa/package.py]: @junghans, v1.0 does not exist, only v0.1, but that is also very old. The tags/releases so far really only mean some kind of development milestone, not really a real release. Can we just link the master branch?

Comment (jbakosi) [inline var/spack/repos/builtin/packages/quinoa/package.py]: Oh, I guess the line below will do just that.

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/quinoa/package.py]: There's a bug in Spack right now that causes problems if you don't have a `url` in your package. Even though this URL will never be used, we need something for now.

Comment (junghans) [inline var/spack/repos/builtin/packages/quinoa/package.py]: The url also help to find future releases, but anyhow it should be `https://github.com/quinoacomputing/quinoa/tarball/Quinoa_v0.1` as we follow the `Quinoa_vX.X` version scheme.

**Comment by junghans** [comment]
There is still some linking error:
```
../LinSys/libquinoa_linsys.so: undefined reference to `Tpetra::CrsMatrix<int, long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokk
os::Serial, Kokkos::HostSpace>, false>::fillComplete(Teuchos::RCP<Tpetra::Map<long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokk
os::Serial, Kokkos::HostSpace> > const> const&, Teuchos::RCP<Tpetra::Map<long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::S
erial, Kokkos::HostSpace> > const> const&, Teuchos::RCP<Teuchos::ParameterList> const&)'
```
(details here [build.txt](https://github.com/LLNL/spack/files/924299/build.txt))

File: var/spack/repos/builtin/packages/quinoa/package.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    url      = "https://github.com/quinoacomputing/quinoa/tarball/Quinoa_v0.1"
+
+    root_cmakelists_dir = '../src'
+
+    version('develop', git='https://github.com/quinoacomputing/quinoa', branch='master')
+
+    variant('debug', default=False, description='Build debug version')
+
+    depends_on('hdf5+mpi')
+    depends_on("charm+mpi+shared~netlrts")
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/quinoa/package.py]: Is `~netlrts` needed here?  In general, using `~` in `depends_on()` can cause concretization errors down the line.  It should only be used here if Quinoa *does not work* with `charm+netlrts`.

Comment (junghans) [inline var/spack/repos/builtin/packages/quinoa/package.py]: The charm only allows the one exclusive network backend and `mpi` as well as `netlrts` are enable by default in spack. Hence a simple `spack install charm` will fail without user intervention, same is true for `  depends_on("charm+mpi")`, which quinoa need for mpi incompatibility.

I guess, what is needed here is a non-boolean build variant for the network backend for charm.

Comment (citibeth) [inline var/spack/repos/builtin/packages/quinoa/package.py]: A package that fails with its default Spack variants is a problem.  Looking at `charm`, I think that the problems should be fixed in that package.

If we had #2386, I would suggest that `+mpi` and `+netlrts` get merged into a single multi-valued variant called `backend`, with values `mpi` or `netlrts` or `verbs`.  Then everything would work pretty smoothly.

Without #2386... can you disable all backends by default in `charm/package.py`?  Can you please also default `shared=True`?  With these changes, you can depend on `charm+mpi`.

But that brings up the question... will `quinoa` work when Charm is built with other backends?  If so, it should just depend on `charm`.  It should not try to forward a backend variant to `charm`, the user should set that directly on `charm` if needed.  If you prefer `charm+mpi`, you can set that in your `packages.yaml` file.  (We're working on other ways to do this too).

Comment (jbakosi) [inline var/spack/repos/builtin/packages/quinoa/package.py]: Quinoa will not, at this point, work without Charm++ built on top of MPI.

Comment (citibeth) [inline var/spack/repos/builtin/packages/quinoa/package.py]: Then I would recommend you change `charm/package.py` to set all backends to `False` and `shared` to `True` by default; and then use `depends_on('charm+mpi')` in `quinoa/package.py`.

File: var/spack/repos/builtin/packages/quinoa/package.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    depends_on("trilinos+exodus")
+    depends_on("boost")
+    depends_on("hypre~internal-superlu")
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/quinoa/package.py]: Here's another tilde to track down the reason "why"?

Comment (junghans) [inline var/spack/repos/builtin/packages/quinoa/package.py]: Standard `trilinos` conflicts with standard `hypre` as both come with superlu. `trilinos` comes with `superlu-dist` and `hypre` comes with `internal-superlu`, so we have to pick one.

**Comment by citibeth** [comment]
I think I'm concluding that while tildes can lead to non-concretizable
specs, getting rid of them all right now is not possible.  With the new
concretizer and multi-valued variants, we will probably want to revisit
this issue later.  It's easy to grep for '~'.

**Comment by junghans** [comment]
I am still getting:
```
../LinSys/libquinoa_linsys.so: undefined reference to `Tpetra::CrsMatrix<int, long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace>, false>::fillComplete(Teuchos::RCP<Tpetra::Map<long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const> const&, Teuchos::RCP<Tpetra::Map<long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const> const&, Teuchos::RCP<Teuchos::ParameterList> const&)'
../LinSys/libquinoa_linsys.so: undefined reference to `Tpetra::CrsMatrix<int, long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace>, false>::fillComplete(Teuchos::RCP<Teuchos::ParameterList> const&)'
../LinSys/libquinoa_linsys.so: undefined reference to `Tpetra::Map<long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::isNodeGlobalElement(long) const'
../LinSys/libquinoa_linsys.so: undefined reference to `Tpetra::Map<long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::Map(unsigned long, Teuchos::ArrayView<long const> const&, long, Teuchos::RCP<Teuchos::Comm<int> const> const&, Teuchos::RCP<Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const&)'
../LinSys/libquinoa_linsys.so: undefined reference to `Teuchos::RCP<Tpetra::Map<long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const> Tpetra::createOneToOne<long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >(Teuchos::RCP<Tpetra::Map<long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const> const&)'
../LinSys/libquinoa_linsys.so: undefined reference to `Tpetra::Vector<double, long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace>, false>::replaceLocalValue(long, double const&) const'
../LinSys/libquinoa_linsys.so: undefined reference to `Tpetra::MultiVector<double, long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace>, false>::getData(unsigned long) const'
../LinSys/libquinoa_linsys.so: undefined reference to `Tpetra::Map<long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::~Map()'
../LinSys/libquinoa_linsys.so: undefined reference to `void Tpetra::MatrixMatrix::Multiply<int, long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >(Tpetra::CrsMatrix<int, long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace>::classic> const&, bool, Tpetra::CrsMatrix<int, long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace>::classic> const&, bool, Tpetra::CrsMatrix<int, long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace>::classic>&, bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Teuchos::RCP<Teuchos::ParameterList> const&)'
../LinSys/libquinoa_linsys.so: undefined reference to `Tpetra::CrsMatrix<int, long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace>, false>::insertGlobalValues(long, Teuchos::ArrayView<long const> const&, Teuchos::ArrayView<int const> const&)'
../LinSys/libquinoa_linsys.so: undefined reference to `Tpetra::Map<long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::getRemoteIndexList(Teuchos::ArrayView<long const> const&, Teuchos::ArrayView<int> const&, Teuchos::ArrayView<long> const&) const'
../LinSys/libquinoa_linsys.so: undefined reference to `Tpetra::Import<long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::Import(Teuchos::RCP<Tpetra::Map<long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const> const&, Teuchos::RCP<Tpetra::Map<long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const> const&)'
../LinSys/libquinoa_linsys.so: undefined reference to `Tpetra::Vector<double, long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace>, false>::Vector(Teuchos::RCP<Tpetra::Map<long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const> const&, bool)'
../LinSys/libquinoa_linsys.so: undefined reference to `Tpetra::CrsMatrix<int, long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace>, false>::CrsMatrix(Teuchos::RCP<Tpetra::Map<long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const> const&, unsigned long, Tpetra::ProfileType, Teuchos::RCP<Teuchos::ParameterList> const&)'
../LinSys/libquinoa_linsys.so: undefined reference to `Tpetra::Vector<double, long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace>, false>::~Vector()'
../LinSys/libquinoa_linsys.so: undefined reference to `Tpetra::DistObject<double, long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace>, false>::doImport(Tpetra::SrcDistObject const&, Tpetra::Import<long, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> > const&, Tpetra::CombineMode)'
collect2: error: ld returned 1 exit status
```
@davydden : have you seen that before?

**Comment by davydden** [comment]
No, never saw that. You can check Trilinos build logs to see if TPetra is ON, but AFAIR it should be.

**Comment by junghans** [comment]
@davydden [here](https://github.com/LLNL/spack/files/955839/cmake.txt) is the cmake output.

**Comment by davydden** [comment]
@junghans output has
```
-- Setting Trilinos_ENABLE_TpetraClassic=ON
-- Setting Trilinos_ENABLE_TpetraKernels=ON
-- Setting Trilinos_ENABLE_TpetraTSQR=ON
-- Setting Trilinos_ENABLE_TpetraCore=ON
-- Setting Trilinos_ENABLE_Tpetra=ON
```
it looks good to me. I would suggest writing to Trilinos mailing list https://trilinos.org/mailman/listinfo/trilinos-users

**Comment by junghans** [comment]
@csiefer2: Have you seen that issue before?

**Comment by csiefer2** [comment]
Nothing obvious looks wrong...

**Comment by junghans** [comment]
This needs #4105 now.

**Comment by junghans** [comment]
@rspavel: Please test this.

**Comment by junghans** [comment]
@citibeth: please review again!

**Pull request closed without merging**